api/communication/communication-core/src/main/java/jakarta/nosql/ServiceLoaderProvider.java

```diff
@@ -60,6 +62,18 @@ public static <T> T get(Class<T> supplier) {
         return getSupplier(requireNonNull(supplier, "supplier is required"));
     }
 
+    /**
+     * Specified a custom service loader mechanism to replace the normal {@link ServiceLoader}-based
+     * behavior. This is intended for environments like OSGi where service loading may have
+     * specialized requirements.
+     * 
+     * @param loader a {@link Function} to produce a stream of service implementations for
+     *               a given class, or {@code null} to un-set an existing custom loader
+     */
+    public static void setLoader(Function<Class<?>, Stream<?>> loader) {
```

Comment:
it is the most dangerous part of the API.
Mainly because it is not safe when someone forgets to set this loader.

I believe this is a good start, but, maybe, start outside the API.